test/framework/src/main/java/org/elasticsearch/test/discovery/TestZenDiscovery.java

```diff
@@ -76,29 +80,40 @@ public TestPlugin(Settings settings) {
                                                                   NamedWriteableRegistry namedWriteableRegistry,
                                                                   MasterService masterService, ClusterApplier clusterApplier,
                                                                   ClusterSettings clusterSettings, UnicastHostsProvider hostsProvider,
-                                                                  AllocationService allocationService) {
+                                                                  AllocationService allocationService, GatewayMetaState gatewayMetaState) {
             // we don't get the latest setting which were updated by the extra settings for the plugin. TODO: fix.
             Settings fixedSettings = Settings.builder().put(settings).putList(DISCOVERY_ZEN_PING_UNICAST_HOSTS_SETTING.getKey()).build();
             return Collections.singletonMap("test-zen", () -> {
                 if (USE_ZEN2.get(settings)) {
-                    // TODO: needs a proper storage layer
-                    Supplier<CoordinationState.PersistedState> persistedStateSupplier =
-                        () -> new InMemoryPersistedState(0L, ClusterState.builder(ClusterName.CLUSTER_NAME_SETTING.get(settings))
-                            .nodes(DiscoveryNodes.builder().add(transportService.getLocalNode())
-                                .localNodeId(transportService.getLocalNode().getId()).build()).build());
+                    Supplier<CoordinationState.PersistedState> persistedStateSupplier;
+                    if (USE_ZEN2_PERSISTED_STATE.get(settings)) {
+                        persistedStateSupplier = () -> {
+                            gatewayMetaState.setLocalNode(transportService.getLocalNode());
+                            return gatewayMetaState;
+                        };
+                    } else {
+                        //if InMemoryPersisted is used, we let GatewayMetaState receive all events
```

Comment:
... because some tests rely on it due to dangling indices functionality.

Reply:
https://github.com/elastic/elasticsearch/pull/35819/commits/d91ad3cab253cf519ec906445e7976c1be26d8d3